src/core/reference/include/openvino/reference/multinomial.hpp

```diff
@@ -0,0 +1,161 @@
+    std::vector<int64_t> target_axis_vec{static_cast<int64_t>(last_axis)};
+    slice(reinterpret_cast<const char*>(cdf.data()),
+          probs_shape,  // == cdf shape
+          reinterpret_cast<char*>(max_value_per_batch.data()),
+          max_value_per_batch_shape,
+          sizeof(T),
+          start,
+          step,
+          target_axis_vec);
+
+    // Normalize the cdf by dividing all elements by the max value in each batch
+    std::vector<T> max_value_per_batch_divisor(total_inputs_elements_count);
+    ov::AxisSet target_axis_set = ov::AxisSet({last_axis});
+    broadcast(reinterpret_cast<const char*>(max_value_per_batch.data()),
+              reinterpret_cast<char*>(max_value_per_batch_divisor.data()),
+              max_value_per_batch_shape,
+              probs_shape,  // expand to original shape (expands last dim)
+              target_axis_set,
+              sizeof(T));
+    divide(cdf.data(), max_value_per_batch_divisor.data(), cdf.data(), total_inputs_elements_count, false);
+
+    // Generate random probability samples
+    std::vector<double> uniform_samples(total_output_elements_count);
+    const double zero = 0;
+    const double one = 1;
+    ov::Shape output_shape_shape{output_shape.size()};
+    std::vector<uint64_t> output_shape_u64(output_shape.begin(), output_shape.end());
+    std::pair<uint64_t, uint64_t> initial_state(0, 0);
+    random_uniform(output_shape_u64.data(),
+                   reinterpret_cast<const char*>(&zero),
+                   reinterpret_cast<const char*>(&one),
+                   reinterpret_cast<char*>(uniform_samples.data()),
+                   output_shape_shape,
+                   ov::element::f64,
+                   global_seed,
+                   op_seed,
+                   initial_state);
+
+    auto first_dim_size = probs_shape.size() == 2 ? probs_shape[0] : (size_t)1;
+    auto second_input_dim_size = probs_shape.size() == 2 ? probs_shape[1] : probs_shape[0];
+    auto second_output_dim_size = probs_shape.size() == 2 ? (size_t)num_samples[0] : probs_shape[0];
+
+    // Iterate over each channel in uniform samples
+    std::vector<U> output_samples(total_output_elements_count);
+    for (size_t i = 0; i < first_dim_size * second_output_dim_size; i += second_output_dim_size) {
+        for (size_t j = 0; j < second_output_dim_size; ++j) {
+            // Iterate over cdf to find the index for a given sample
+            // If no class found (all have 0 probability), selects last - undefined behavior
+            auto i_translated = i / second_output_dim_size * second_input_dim_size;
+            auto selected_class_idx = second_input_dim_size;
+            auto sample_value = uniform_samples[i + j];
+            for (size_t l = 0; l < second_input_dim_size; ++l) {
+                if (sample_value <= cdf[i_translated + l]) {
+                    output_samples[i + j] = l;  // warn
+                    selected_class_idx = l;
+                    break;
```

Comment:
It's better to avoid `l` (small L) as a variable name, as it's easy to mislead it with `1` (number one).
What is the `// warn` comment about?
```suggestion
            for (size_t class_idx = 0; class_idx < class_size; ++class_idx) {
                if (sample_value <= cdf[i_translated + class_idx]) {
                    output_samples[i + j] = class_idx;  // warn
                    selected_class_idx = class_idx;
                    break;
```

Reply:
Changed to `k`! This warning is to verify whether the class_idx (size_t) can be safely cast to type of output_samples (U), but since they are primitives I now removed this warning